cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -1384,9 +1385,255 @@ size_t CacheAllocator<CacheTrait>::wakeUpWaitersLocked(folly::StringPiece key,
+    case dml::status_code::queue_busy:
+      /* Enqueue failed to one or several queues */
+      error = "busy";
+      break;
+    case dml::status_code::error:
+      /* Internal library error occurred */
+      error = "internal";
+      break;
+    case dml::status_code::ok:
+      /* should not be here */
+      error = "ok";
+      break;
+    default:
+      error = "none of the above - okay!!";
+      break;
+  }
+  return error + " mem copy error";
+}
+
+using allocator_t = std::allocator<dml::byte_t>;
+using batch_handler_t = dml::handler<dml::batch_operation, allocator_t>;
+
+template <typename CacheTrait>
+void CacheAllocator<CacheTrait>::evictRegularItems(TierId tid, PoolId pid, ClassId cid,
+                                                   std::vector<EvictionData>& evictionData,
+                                                   std::vector<WriteHandle>& newItemHdls,
+                                                   bool skipAddInMMContainer,
+                                                   bool fromBgThread,
+                                                   std::vector<bool>& moved) {
+  /* Split batch for DSA-based move */
+  const auto& pool = allocator_[tid]->getPool(pid);
+  const auto& allocSizes = pool.getAllocSizes();
+  auto isLarge = allocSizes[cid] >= config_.largeItemMinSize;
+  auto dmlBatchRatio = isLarge ? config_.largeItemBatchEvictDsaUsageFraction :
+                                 config_.smallItemBatchEvictDsaUsageFraction;
+  size_t dmlBatchSize =
+      (config_.dsaEnabled && evictionData.size() >= config_.minBatchSizeForDsaUsage) ?
+                    static_cast<size_t>(evictionData.size() * dmlBatchRatio) : 0;
+  auto sequence = dml::sequence<allocator_t>(dmlBatchSize);
+  batch_handler_t handler{};
+
+  /* Move a calculated portion of the batch using DSA (if enabled) */
+  for (auto i = 0U; i < dmlBatchSize; i++) {
+    XDCHECK(!evictionData[i].candidate->isExpired());
+    XDCHECK_EQ(newItemHdls[i]->getSize(), evictionData[i].candidate->getSize());
+    if (evictionData[i].candidate->isNvmClean()) {
+      newItemHdls[i]->markNvmClean();
+    }
+    dml::const_data_view srcView = dml::make_view(
+          reinterpret_cast<uint8_t*>(evictionData[i].candidate->getMemory()),
+          evictionData[i].candidate->getSize());
+    dml::data_view dstView = dml::make_view(
+          reinterpret_cast<uint8_t*>(newItemHdls[i]->getMemory()),
+          newItemHdls[i]->getSize());
+    if (sequence.add(dml::mem_copy, srcView, dstView) != dml::status_code::ok) {
+      throw std::runtime_error(folly::sformat(
+          "failed to add dml::mem_copy operation to the sequence for item: {}",
+          evictionData[i].candidate->toString()));
+    }
+  }
+  if (dmlBatchSize) {
+    handler = dml::submit<dml::hardware>(dml::batch, sequence);
+    if (!handler.valid()) {
+      auto status = handler.get();
+      XDCHECK(handler.valid()) << dmlErrStr(status);
+      throw std::runtime_error(folly::sformat(
```

Comment:
How often do we expect this to happen? I'm wondering if it would make sense to return an error code here instead of throwing an exception

Reply:
I have not seen this error if DML is configured correctly. Theoretically it is possible to hit this error if we oversubscribe to DSA work queues. But that's more of a config issue - the DSA vs CPU batch split adjustment has to be tuned to prevent this scenario. If, in the near future, I add any self-tuned mechanism for this DSA vs CPU batch split, then it makes sense to use this error scenario to adjust the ratio and proceed rather than throw a runtime exception.